src/main/java/slimeknights/tconstruct/library/modifiers/modules/technical/MaxArmorLevelModule.java

```diff
@@ -0,0 +1,119 @@
+    MaxArmorLevelModule::new);
+
+  @Override
+  public RecordLoadable<MaxArmorLevelModule> getLoader() {
+    return LOADER;
+  }
+
+  @Override
+  public List<ModuleHook<?>> getDefaultHooks() {
+    return DEFAULT_HOOKS;
+  }
+
+  @Override
+  public void onEquip(IToolStackView tool, ModifierEntry modifier, EquipmentChangeContext context) {
+    findMaxLevelIfArmor(tool, context, key, modifier.intEffectiveLevel(), allowBroken, heldTag);
+  }
+
+  @Override
+  public void onUnequip(IToolStackView tool, ModifierEntry modifier, EquipmentChangeContext context) {
+    findMaxLevelIfArmor(tool, context, key, -modifier.intEffectiveLevel(), allowBroken, heldTag);
+  }
+
+
+  /* Helpers */
+
+  /**
+   * Adds levels to the given key in entity modifier data for an armor modifier
+   * @param context  Equipment change context
+   * @param key      Key to modify
+   * @param amount   Amount to add
+   */
+  public static void findMaxLevel(EquipmentChangeContext context, TinkerDataKey<Integer> key, int amount) {
+    context.getTinkerData().ifPresent(data -> {
+      int maxLevel = Math.max(data.get(key, 0), amount);
+      if (maxLevel <= 0) {
+        data.remove(key);
+      } else {
+        data.put(key, maxLevel);
+      }
+    });
+  }
+
+  /** Checks if the given slot is valid */
+  public static boolean validSlot(IToolStackView tool, EquipmentSlot slot, @Nullable TagKey<Item> heldTag) {
+    return slot.getType() == Type.ARMOR || heldTag != null && tool.hasTag(heldTag);
+  }
+
+  /**
+   * Adds levels to the given key in entity modifier data for an armor modifier
+   * @param tool     Tool instance
+   * @param context  Equipment change context
+   * @param key      Key to modify
+   * @param amount   Amount to add
+   * @param heldTag  Tag to check to validate held items, if null held items are considered to never be valid
+   */
+  public static void findMaxLevelIfArmor(IToolStackView tool, EquipmentChangeContext context, TinkerDataKey<Integer> key, int amount, boolean allowBroken, @Nullable TagKey<Item> heldTag) {
+    if (validSlot(tool, context.getChangedSlot(), heldTag) && (allowBroken || !tool.isBroken())) {
+      findMaxLevel(context, key, amount);
+    }
+  }
+
+  /**
+   * Gets the total level from the key in the entity modifier data
+   * @param living  Living entity
+   * @param key     Key to get
+   * @return  Level from the key
+   */
+  public static int getLevel(LivingEntity living, TinkerDataKey<Integer> key) {
+    return getLevel(living.getCapability(TinkerDataCapability.CAPABILITY), key);
+  }
+
+  /**
+   * Gets the total level from the key in the entity modifier data
+   * @param cap    Capability instance
+   * @param key    Key to get
+   * @return  Level from the key
+   */
+  public static int getLevel(LazyOptional<TinkerDataCapability.Holder> cap, TinkerDataKey<Integer> key) {
+    return cap.resolve().map(data -> data.get(key)).orElse(0);
+  }
```

Comment:
These seem redundant to those in `ArmorLevelModule`, I'd just ditch the extra copies.